configfile: "snakemake_config.yaml"

BCALM = config["bcalm_bin_path"]
BCALMCONVERTPATH = config["bcalm_convert_path"]
ART = config["art_bin_path"]
PBSIM = config["pbsim_bin_path"]
RANDOMREADS = config["randomreads_path"]

ASTARIX_BIN = config["astarix_bin_path"]
GA_BIN = config["graphaligner_bin_path"]
PASGAL_BIN = config["pasgal_bin_path"]
VG_BIN = config["vg_bin_path"]
VARGAS_BIN = config["vargas_bin_path"]

TIMELIMIT       = "(timeout 3600s"
TIMEOUTPRINT    = ") || ([ $? -eq 124 ] && echo -e \"timeout\" >{log}) || (echo -e \"exit error code\n$?\" | tee {log})"

DIJKSTRA_CMD								= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a dijkstra     -D 14 -M 0 -S 1 -G 5                                                             -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_PREFIX_ILLUMINA_CMD					= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-prefix -D 14 -M 0 -S 1 -G 5 -d 5                                                        -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_PREFIX_CMD							= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-prefix -D 14 -M 0 -S 1 -G 1 -d 5                                                        -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_ILLUMINA_CMD 					= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 5 --seeds_len {wildcards.seedlen}                                  -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
#ASTARIX_SEEDS_EXACT_SHORT_CMD 				= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen}                                  -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_EXACT_CMD 					= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen}                                  -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_EXACT_WO_MATCH_POS_CMD		= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen} --seeds_match_pos_optimization 0 -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_EXACT_WO_SKIP_NEAR_CRUMBS_CMD = "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen} --seeds_skip_near_crumbs 0       -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_APPROX_CMD					= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen} --seeds_max_errors 1             -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"
ASTARIX_SEEDS_EXACT_WO_OPT_CMD				= "{TIMELIMIT} {ASTARIX_BIN} align-optimal -a astar-seeds  -D 14 -M 0 -S 1 -G 1 --seeds_len {wildcards.seedlen} --seeds_match_pos_optimization 0 --seeds_skip_near_crumbs 0 -q {input.reads} -g $(echo {input.graph} | sed 's|_head100000000|_head0|g') -o $(dirname {output.align}) >{output.file} {TIMEOUTPRINT} >>{output.file}"

GA_CMD = "{TIMELIMIT} {GA_BIN} --seeds-first-full-rows 64 -b 10000 -t 1 -a alignments.gaf -f {input.reads} -g {input.graph_vg} >{output.file} {TIMEOUTPRINT} >>{output.file}"
PASGAL_CMD = "{TIMELIMIT} {PASGAL_BIN} -m vg -r {input.graph_vg} -q {input.reads} -t 1 -o {output.file} >{output.file} {TIMEOUTPRINT}"
VARGAS_CMD = "{TIMELIMIT} {VARGAS_BIN} align -g {input.graph_gdef} -U {input.reads} --ete >{output.align} 2>{output.file} {TIMEOUTPRINT}"

PaSGAL_DATA_URL = "https://alurulab.cc.gatech.edu/sites/all/projectFiles/PaSGAL_Chirag"
PaSGAL_FILES = ["L1.fa", "L2.fastq", "L3.fastq",
                "M1.fa", "M2.fastq", "M3.fastq",
                "LRC.vg", "MHC1.vg", "MHC2.txt"]

# LINEAR GENOMES USED FOR:
# -- PaSGAL does not accept .gaf and bi-directional .vg;
# -- Vargas does not accept .gaf and .vg but only .gdef that can be build by .fa

TABLE_ILLUMINA_ALGOS          = ["astarix-seeds-illumina", "graphaligner", "pasgal", "vargas", "astarix-prefix-illumina", "dijkstra"]
TABLE_ILLUMINA_Ns             = [100000, 1000, 1000, 1000, 1000, 1000]  # number of reads
TABLE_ILLUMINA_GRAPHS         = ["MHC1", "ecoli"] #, "LRC"]
TABLE_ILLUMINA_GRAPH_HEADS    = [100000000]
TABLE_ILLUMINA_SEQ_TECHNOLOGY = ["illumina"]
TABLE_ILLUMINA_ms             = [200]
TABLE_ILLUMINA_errors         = [-1]
TABLE_ILLUMINA_seedlen        = [25]

TABLE_HIFI_ALGOS          = ["astarix-seeds", "graphaligner", "vargas", "pasgal", "dijkstra", "astarix-prefix"]
TABLE_HIFI_Ns             = [2, 0.1, 0.02, 0.1, 0.1, 0.1]  # coverage of reads
TABLE_HIFI_GRAPHS         = ["MHC1", "ecoli"]
TABLE_HIFI_GRAPH_HEADS    = [100000000]
TABLE_HIFI_SEQ_TECHNOLOGY = ["hifi-natural"]
TABLE_HIFI_ms             = [0]
TABLE_HIFI_errors         = [0.003]
TABLE_HIFI_seedlen        = [50]

REFSIZE_ILLUMINA_ALGOS          = ["astarix-seeds-illumina", "vargas", "graphaligner", "pasgal", "astarix-prefix-illumina", "dijkstra"]
REFSIZE_ILLUMINA_Ns             = [10000, 1000, 1000, 1000, 1000, 1000]  # number of reads
REFSIZE_ILLUMINA_GRAPHS         = ["MHC1"]
REFSIZE_ILLUMINA_GRAPH_HEADS    = [5000, 10000, 20000, 50000, 500000, 5000000]
REFSIZE_ILLUMINA_SEQ_TECHNOLOGY = ["illumina"]
REFSIZE_ILLUMINA_ms             = [200]
REFSIZE_ILLUMINA_errors         = [-1]
REFSIZE_ILLUMINA_seedlen        = [25]

REFSIZE_HIFI_ALGOS          = ["astarix-seeds", "graphaligner", "pasgal", "vargas"] #, "astarix-prefix-illumina", ] # "dijkstra"
REFSIZE_HIFI_Ns             = [1, 0.1, 0.02, 0.02]  # 0.02 coverage of reads
REFSIZE_HIFI_GRAPHS         = ["MHC1"]
REFSIZE_HIFI_GRAPH_HEADS    = [50000, 500000, 5000000]
REFSIZE_HIFI_SEQ_TECHNOLOGY = ["hifi-natural"]
REFSIZE_HIFI_ms             = ["-1"]
REFSIZE_HIFI_errors         = [0.003]
REFSIZE_HIFI_seedlen        = [50]

READLEN_ALGOS           = ["astarix-seeds", "graphaligner", "pasgal", "vargas"]
READLEN_Ns              = [1, 0.02, 0.02, 0.02]
READLEN_GRAPHS          = ["MHC1"]  #["ecoli"]
READLEN_GRAPH_HEADS     = [10000000]  # 0 stands for the .gfa graph version, not the linear .fa
READLEN_SEQ_TECHNOLOGY  = ["hifi"] #"ccs"
READLEN_ms              = [2000, 7000, 15000, 25000]
READLEN_errors          = [0.003]
READLEN_seedlen         = [50]

# NOT USED
ABLATION_ALGOS          = ["astarix-seeds", "astarix-seeds_wo_skip_near_crumbs_pos"] #"astarix-seeds_wo_match_pos"] #"astarix-seeds_wo_opt", "vargas"]
ABLATION_Ns             = [1, 0.1]  # coverage
ABLATION_GRAPHS         = ["MHC1"] #["ecoli"]
ABLATION_GRAPH_HEADS    = [100000000]
ABLATION_SEQ_TECHNOLOGY = ["hifi"] #"ccs", "illumina",
ABLATION_ms             = [2000, 10000, 20000, 25000] #, 25000, 30000] # [100, 300, 700, 1000, 2000] #, 3000, 5000, 7000, 10000, 13000, 15000, 18000, 20000]
ABLATION_errors         = [0.003]
ABLATION_seedlen        = [50]

# NOT USED
ERRORRATE_ALGOS          = ["astarix-seeds-illumina"] #, "astarix-seeds_wo_match_pos"]
ERRORRATE_GRAPHS         = ["MHC1"]
ERRORRATE_GRAPH_HEADS    = [0]
ERRORRATE_Ns             = [1000, 10000, 50000, 100000]
ERRORRATE_SEQ_TECHNOLOGY = ["illumina"] #"ccs", "hifi",
ERRORRATE_ms             = [200] # [100, 300, 700, 1000, 2000] #, 3000, 5000, 7000, 10000, 13000, 15000, 18000, 20000]
ERRORRATE_errors         = [-1] # 0.005, 0.01, 0.02, 0.03
ERRORRATE_seedlen        = [25]

FIRST_ROW       = "algo\tgraph\thead\tN\tm\tsequencing_technology\terrorrate\tseedlen\ts\th:m:s\tmax_rss\tmax_vms\tmax_uss\tmax_pss\tio_in\tio_out\tmean_load"

graph_params = "{graph}_head{head}"
read_params  = "N{N}_m{m}_{seq_technology}_errors{errors}"
params       = graph_params + "_" + read_params + "_seedlen{seedlen}"
prefix       = "{algo}\t{graph}\t{head}\t{N}\t{m}\t{seq_technology}\t{errors}\t{seedlen}"

graph_params1 = "{{graph}}_head{{head}}"
read_params1  = "N{N}_m{{m}}_{{seq_technology}}_errors{{errors}}"
params1       = graph_params1 + "_" + read_params1 + "_seedlen{{seedlen}}"
prefix1       = "{algo}\t{{graph}}\t{{head}}\t{N}\t{{m}}\t{{seq_technology}}\t{{errors}}\t{{seedlen}}"

#EXPERIMENT_N = 1000 #20000
#EXPERIMENT_FIRST_ROW       = "algo\tgraph\ts\th:m:s\tmax_rss\tmax_vms\tmax_uss\tmax_pss\tio_in\tio_out\tmean_load"
#EXPERIMENT_GRAPHS = ["MHC1", "LRC"]
#EXPERIMENT_READS = { "MHC1": "M1", "LRC": "L1", "ecoli": "ecoli"}

#experiment_graph = "graphs/pasgal-{graph}.gfa"
#experiment_graph_vg = "raw/PaSGAL/{graph}.vg"
#experiment_input_reads = "raw/PaSGAL/{reads}.fa"

aligners_inputs = {
    'graph' : "graphs/"+graph_params+".gfa",
    'graph_vg' : "graphs/"+graph_params+".vg",
    'reads' : "reads/"+params+"/reads.fq"
}

## Default summary rule to be satisfied.
rule all:
    input:
#        agg1 = "results/aggregated_benchmarks.csv"
        table_illumina = "results/TABLE_ILLUMINA",
        table_hifi = "results/TABLE_HIFI",
        ref_size_illumina = "results/REFSIZE_ILLUMINA",
        ref_size_hifi = "results/REFSIZE_HIFI",
        readlen = "results/READLEN",
#        ablation = "results/ABLATION",
#        errorrate = "results/ERRORRATE",
#        numreads = "results/NUMREADS"
#        astar_prefix_vs_seeds = "results/PREFIX_VS_SEEDS_RAN",
#        visualization = "results/VISUALIZATION_RAN",
#        agg2 = "results/aggregated_experiments.csv",
#        trie = "results/TRIE_DUMMY"

table_illumina_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(TABLE_ILLUMINA_ALGOS, TABLE_ILLUMINA_Ns)]
table_illumina_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(TABLE_ILLUMINA_ALGOS, TABLE_ILLUMINA_Ns)]
table_illumina_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(TABLE_ILLUMINA_ALGOS, TABLE_ILLUMINA_Ns)]
rule table_illumina:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=TABLE_ILLUMINA_GRAPHS, head=TABLE_ILLUMINA_GRAPH_HEADS),
        reads = expand(table_illumina_zippedreads, graph=TABLE_ILLUMINA_GRAPHS, head=TABLE_ILLUMINA_GRAPH_HEADS, m=TABLE_ILLUMINA_ms, seq_technology=TABLE_ILLUMINA_SEQ_TECHNOLOGY, errors=TABLE_ILLUMINA_errors, seedlen=TABLE_ILLUMINA_seedlen),
        file = expand(table_illumina_zippedfile, graph=TABLE_ILLUMINA_GRAPHS, head=TABLE_ILLUMINA_GRAPH_HEADS, m=TABLE_ILLUMINA_ms, seq_technology=TABLE_ILLUMINA_SEQ_TECHNOLOGY, errors=TABLE_ILLUMINA_errors, seedlen=TABLE_ILLUMINA_seedlen),
    output:
        "results/TABLE_ILLUMINA"
    params:
        prefix = expand(table_illumina_zippedprefix, graph=TABLE_ILLUMINA_GRAPHS, head=TABLE_ILLUMINA_GRAPH_HEADS, N=TABLE_ILLUMINA_Ns, m=TABLE_ILLUMINA_ms, seq_technology=TABLE_ILLUMINA_SEQ_TECHNOLOGY, errors=TABLE_ILLUMINA_errors, seedlen=TABLE_ILLUMINA_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

table_hifi_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(TABLE_HIFI_ALGOS, TABLE_HIFI_Ns)]
table_hifi_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(TABLE_HIFI_ALGOS, TABLE_HIFI_Ns)]
table_hifi_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(TABLE_HIFI_ALGOS, TABLE_HIFI_Ns)]
rule table_hifi:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=TABLE_HIFI_GRAPHS, head=TABLE_HIFI_GRAPH_HEADS),
        reads = expand(table_hifi_zippedreads, graph=TABLE_HIFI_GRAPHS, head=TABLE_HIFI_GRAPH_HEADS, m=TABLE_HIFI_ms, seq_technology=TABLE_HIFI_SEQ_TECHNOLOGY, errors=TABLE_HIFI_errors, seedlen=TABLE_HIFI_seedlen),
        file = expand(table_hifi_zippedfile, graph=TABLE_HIFI_GRAPHS, head=TABLE_HIFI_GRAPH_HEADS, m=TABLE_HIFI_ms, seq_technology=TABLE_HIFI_SEQ_TECHNOLOGY, errors=TABLE_HIFI_errors, seedlen=TABLE_HIFI_seedlen),
    output:
        "results/TABLE_HIFI"
    params:
        prefix = expand(table_hifi_zippedprefix, graph=TABLE_HIFI_GRAPHS, head=TABLE_HIFI_GRAPH_HEADS, N=TABLE_HIFI_Ns, m=TABLE_HIFI_ms, seq_technology=TABLE_HIFI_SEQ_TECHNOLOGY, errors=TABLE_HIFI_errors, seedlen=TABLE_HIFI_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

refsize_illumina_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_ILLUMINA_ALGOS, REFSIZE_ILLUMINA_Ns)]
refsize_illumina_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_ILLUMINA_ALGOS, REFSIZE_ILLUMINA_Ns)]
refsize_illumina_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_ILLUMINA_ALGOS, REFSIZE_ILLUMINA_Ns)]
rule refsize_illumina:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=REFSIZE_ILLUMINA_GRAPHS, head=REFSIZE_ILLUMINA_GRAPH_HEADS),
        reads = expand(refsize_illumina_zippedreads, graph=REFSIZE_ILLUMINA_GRAPHS, head=REFSIZE_ILLUMINA_GRAPH_HEADS, m=REFSIZE_ILLUMINA_ms, seq_technology=REFSIZE_ILLUMINA_SEQ_TECHNOLOGY, errors=REFSIZE_ILLUMINA_errors, seedlen=REFSIZE_ILLUMINA_seedlen),
        file = expand(refsize_illumina_zippedfile, graph=REFSIZE_ILLUMINA_GRAPHS, head=REFSIZE_ILLUMINA_GRAPH_HEADS, m=REFSIZE_ILLUMINA_ms, seq_technology=REFSIZE_ILLUMINA_SEQ_TECHNOLOGY, errors=REFSIZE_ILLUMINA_errors, seedlen=REFSIZE_ILLUMINA_seedlen),
    output:
        "results/REFSIZE_ILLUMINA"
    params:
        prefix = expand(refsize_illumina_zippedprefix, graph=REFSIZE_ILLUMINA_GRAPHS, head=REFSIZE_ILLUMINA_GRAPH_HEADS, N=REFSIZE_ILLUMINA_Ns, m=REFSIZE_ILLUMINA_ms, seq_technology=REFSIZE_ILLUMINA_SEQ_TECHNOLOGY, errors=REFSIZE_ILLUMINA_errors, seedlen=REFSIZE_ILLUMINA_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

refsize_hifi_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_HIFI_ALGOS, REFSIZE_HIFI_Ns)]
refsize_hifi_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_HIFI_ALGOS, REFSIZE_HIFI_Ns)]
refsize_hifi_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(REFSIZE_HIFI_ALGOS, REFSIZE_HIFI_Ns)]
rule refsize_hifi:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=REFSIZE_HIFI_GRAPHS, head=REFSIZE_HIFI_GRAPH_HEADS),
        reads = expand(refsize_hifi_zippedreads, graph=REFSIZE_HIFI_GRAPHS, head=REFSIZE_HIFI_GRAPH_HEADS, m=REFSIZE_HIFI_ms, seq_technology=REFSIZE_HIFI_SEQ_TECHNOLOGY, errors=REFSIZE_HIFI_errors, seedlen=REFSIZE_HIFI_seedlen),
        file = expand(refsize_hifi_zippedfile, graph=REFSIZE_HIFI_GRAPHS, head=REFSIZE_HIFI_GRAPH_HEADS, m=REFSIZE_HIFI_ms, seq_technology=REFSIZE_HIFI_SEQ_TECHNOLOGY, errors=REFSIZE_HIFI_errors, seedlen=REFSIZE_HIFI_seedlen),
    output:
        "results/REFSIZE_HIFI"
    params:
        prefix = expand(refsize_hifi_zippedprefix, graph=REFSIZE_HIFI_GRAPHS, head=REFSIZE_HIFI_GRAPH_HEADS, N=REFSIZE_HIFI_Ns, m=REFSIZE_HIFI_ms, seq_technology=REFSIZE_HIFI_SEQ_TECHNOLOGY, errors=REFSIZE_HIFI_errors, seedlen=REFSIZE_HIFI_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

readlen_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(READLEN_ALGOS, READLEN_Ns)]
readlen_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(READLEN_ALGOS, READLEN_Ns)]
readlen_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(READLEN_ALGOS, READLEN_Ns)]
rule readlen:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=READLEN_GRAPHS, head=READLEN_GRAPH_HEADS),
        reads = expand(readlen_zippedreads, graph=READLEN_GRAPHS, head=READLEN_GRAPH_HEADS, m=READLEN_ms, seq_technology=READLEN_SEQ_TECHNOLOGY, errors=READLEN_errors, seedlen=READLEN_seedlen),
        file = expand(readlen_zippedfile, graph=READLEN_GRAPHS, head=READLEN_GRAPH_HEADS, m=READLEN_ms, seq_technology=READLEN_SEQ_TECHNOLOGY, errors=READLEN_errors, seedlen=READLEN_seedlen),
    output:
        "results/READLEN"
    params:
        prefix = expand(readlen_zippedprefix, graph=READLEN_GRAPHS, head=READLEN_GRAPH_HEADS, N=READLEN_Ns, m=READLEN_ms, seq_technology=READLEN_SEQ_TECHNOLOGY, errors=READLEN_errors, seedlen=READLEN_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

ablation_zippedreads = [("reads/"+params1+"/reads.fq").format(algo=algo, N=N) for (algo, N) in zip(ABLATION_ALGOS, ABLATION_Ns)]
ablation_zippedfile = [("results/{algo}/"+params1+"/benchmark.txt").format(algo=algo, N=N) for (algo, N) in zip(ABLATION_ALGOS, ABLATION_Ns)]
ablation_zippedprefix = [prefix1.format(algo=algo, N=N) for (algo, N) in zip(ABLATION_ALGOS, ABLATION_Ns)]
rule ablation:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=ABLATION_GRAPHS, head=ABLATION_GRAPH_HEADS),
        reads = expand(ablation_zippedreads, graph=ABLATION_GRAPHS, head=ABLATION_GRAPH_HEADS, m=ABLATION_ms, seq_technology=ABLATION_SEQ_TECHNOLOGY, errors=ABLATION_errors, seedlen=ABLATION_seedlen),
        file = expand(ablation_zippedfile, graph=ABLATION_GRAPHS, head=ABLATION_GRAPH_HEADS, m=ABLATION_ms, seq_technology=ABLATION_SEQ_TECHNOLOGY, errors=ABLATION_errors, seedlen=ABLATION_seedlen),
    output:
        "results/ABLATION"
    params:
        prefix = expand(ablation_zippedprefix, graph=ABLATION_GRAPHS, head=ABLATION_GRAPH_HEADS, N=ABLATION_Ns, m=ABLATION_ms, seq_technology=ABLATION_SEQ_TECHNOLOGY, errors=ABLATION_errors, seedlen=ABLATION_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

rule errorrate:
    input:
        refs = expand("graphs/"+graph_params+".fa", graph=ERRORRATE_GRAPHS, head=ERRORRATE_GRAPH_HEADS),
        reads = expand("reads/"+params+"/reads.fq", algo=ERRORRATE_ALGOS, graph=ERRORRATE_GRAPHS, head=ERRORRATE_GRAPH_HEADS, N=ERRORRATE_Ns, m=ERRORRATE_ms, seq_technology=ERRORRATE_SEQ_TECHNOLOGY, errors=ERRORRATE_errors, seedlen=ERRORRATE_seedlen),
        file = expand("results/{algo}/"+params+"/benchmark.txt", algo=ERRORRATE_ALGOS, graph=ERRORRATE_GRAPHS, head=ERRORRATE_GRAPH_HEADS, N=ERRORRATE_Ns, m=ERRORRATE_ms, seq_technology=ERRORRATE_SEQ_TECHNOLOGY, errors=ERRORRATE_errors, seedlen=ERRORRATE_seedlen),
    output:
        "results/ERRORRATE"
    params:
        prefix = expand(prefix, algo=ERRORRATE_ALGOS, graph=ERRORRATE_GRAPHS, head=ERRORRATE_GRAPH_HEADS, N=ERRORRATE_Ns, m=ERRORRATE_ms, seq_technology=ERRORRATE_SEQ_TECHNOLOGY, errors=ERRORRATE_errors, seedlen=ERRORRATE_seedlen)
    shell:
        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"
#        "jupyter nbconvert --to notebook --inplace --execute ref_size_scaling.ipynb >{output}"

#rule aggregate_benchmarks:
#    input:
#        file = expand("results/{algo}/"+params+"/benchmark.txt", algo=ALGOS, graph=GRAPHS, head=GRAPH_HEADS, N=Ns, m=ms),
#    output:
#        "results/aggregated_benchmarks.csv"
#    params:
#        prefix = expand("{algo}\t{graph}\t{head}\t{N}\t{m}", algo=ALGOS, graph=GRAPHS, head=GRAPH_HEADS, N=Ns, m=ms)
#    shell:
#        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{FIRST_ROW}\\n/' > {output}"

#rule aggregate_exp:
#    input:
#        file = expand("results/{algo}/{graph}/benchmark.txt", algo=ALGOS, graph=EXPERIMENT_GRAPHS),
#    output:
#        "results/aggregated_experiments.csv"
#    params:
#        prefix = expand("{algo}\t{graph}", algo=ALGOS, graph=EXPERIMENT_GRAPHS)
#    shell:
#        "paste <(echo \"{params.prefix}\" | tr ' ' '\n') <(tail -n 1 --silent {input.file}) | sed '1s/^/{EXPERIMENT_FIRST_ROW}\\n/' > {output}"

#rule visualization_notebook:
#    input:
#        aggregated_benchmarks = "results/aggregated_benchmarks.csv",
#    output:
#        "results/VISUALIZATION_RAN"
#    shell:
#        "jupyter nbconvert --to notebook --inplace --execute visualization.ipynb > {output}"

rule astar_prefix_vs_seeds:
    input:
        astar_prefix_align = "results/astarix-prefix/MHC1/out/alignments.tsv",
        astar_seeds_align = "results/astarix-seeds/MHC1/out/alignments.tsv"
    output:
        "results/PREFIX_VS_SEEDS_RAN"
    shell:
        "jupyter nbconvert --to notebook --inplace --execute astar_vs_dijkstra.ipynb > {output}"

rule astarix_prefix:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-prefix/"+params+"/out/alignments.tsv",
        file = "results/astarix-prefix/"+params+"/summary.txt"
    benchmark:
        "results/astarix-prefix/"+params+"/benchmark.txt"
    log:
        "results/astarix-prefix/"+params+"/log.txt"
    shell:
        ASTARIX_PREFIX_CMD

rule astarix_prefix_illumina:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-prefix-illumina/"+params+"/out/alignments.tsv",
        file = "results/astarix-prefix-illumina/"+params+"/summary.txt"
    benchmark:
        "results/astarix-prefix-illumina/"+params+"/benchmark.txt"
    log:
        "results/astarix-prefix-illumina/"+params+"/log.txt"
    shell:
        ASTARIX_PREFIX_ILLUMINA_CMD

rule astarix_seeds_exact:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_EXACT_CMD

rule astarix_seeds_illumina:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds-illumina/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds-illumina/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds-illumina/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds-illumina/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_ILLUMINA_CMD

rule astarix_seeds_wo_match_pos:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds_wo_match_pos/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds_wo_match_pos/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds_wo_match_pos/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds_wo_match_pos/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_EXACT_WO_MATCH_POS_CMD

rule astarix_seeds_wo_skip_near_crumbs_pos:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds_wo_skip_near_crumbs_pos/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds_wo_skip_near_crumbs_pos/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds_wo_skip_near_crumbs_pos/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds_wo_skip_near_crumbs_pos/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_EXACT_WO_SKIP_NEAR_CRUMBS_CMD

rule astarix_seeds_wo_opt:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds_wo_opt/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds_wo_opt/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds_wo_opt/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds_wo_opt/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_EXACT_WO_OPT_CMD

rule astarix_seeds_approx:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/astarix-seeds-approx/"+params+"/out/alignments.tsv",
        file = "results/astarix-seeds-approx/"+params+"/summary.txt"
    benchmark:
        "results/astarix-seeds-approx/"+params+"/benchmark.txt"
    log:
        "results/astarix-seeds-approx/"+params+"/log.txt"
    shell:
        ASTARIX_SEEDS_APPROX_CMD

rule dijkstra:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph = "graphs/"+graph_params+".gfa"
    output:
        align = "results/dijkstra/"+params+"/out/alignments.tsv",
        file = "results/dijkstra/"+params+"/summary.txt"
    benchmark:
        "results/dijkstra/"+params+"/benchmark.txt"
    log:
        "results/dijkstra/"+params+"/log.txt"
    shell:
        DIJKSTRA_CMD

rule vargas:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph_gdef = "graphs/"+graph_params+".gdef"
    output:
        align = "results/vargas/"+params+"/out/alignments.tsv",
        file = "results/vargas/"+params+"/summary.txt"
    benchmark:
        "results/vargas/"+params+"/benchmark.txt"
    log:
        "results/vargas/"+params+"/log.txt"
    shell:
        VARGAS_CMD

rule prepare_vargas_graph:
    input:
        "graphs/"+graph_params+".fa"
    output:
        "graphs/"+graph_params+".gdef"
    shell:
        "{VARGAS_BIN} define -f {input} -t {output}"

rule graphaligner:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph_vg = "graphs/"+graph_params+".vg"
    output:
        file = "results/graphaligner/"+params+"/summary.txt"
    benchmark:
        "results/graphaligner/"+params+"/benchmark.txt"
    log:
        "results/graphaligner/"+params+"/log.txt"
    shell:
        GA_CMD

rule pasgal:
    input:
        reads = "reads/"+params+"/reads.fq",
        graph_vg = "graphs/"+graph_params+".vg"
    output:
        file = "results/pasgal/"+params+"/summary.txt"
    benchmark:
        "results/pasgal/"+params+"/benchmark.txt"
    log:
        "results/pasgal/"+params+"/log.txt"
    shell:
        PASGAL_CMD

rule vg:
	input: **aligners_inputs
	output:
		file = "results/vg/"+params+"/summary.txt",
		align = "results/vg/"+params+"/alignments.tsv"
	benchmark:
		"results/vg/"+params+"/benchmark.txt"
	shell:
		#{TIMELIMIT} 
		"while mapfile -t -n 4 ary && ((${{#ary[@]}})); do {VG_BIN} align -s \"${{ary[1]}}\" {input.graph_vg} -j; done <{input.reads} >{output.align} 2>{output.file} {TIMEOUTPRINT}"

rule vg2gfa:
    input:
        "raw/PaSGAL/{prefix}.vg"
    output:
        "graphs/pasgal-{prefix}.gfa"
    shell:
        "{VG_BIN} view {input} > {output}"

## Chronological order of execution follows. ##

rule clean:
    shell:
        "rm -rf reads graphs graphs"

rule download_ecoli:
    output:
        protected("raw/ecoli.fa")
    shadow: "shallow"
    run:
        shell("wget ftp://ftp.ensemblgenomes.org/pub/bacteria/release-40/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/dna/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna.chromosome.Chromosome.fa.gz"),
        shell("gunzip Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna.chromosome.Chromosome.fa.gz")
        shell("cut -d ' ' -f 1 < Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna.chromosome.Chromosome.fa >{output}")

rule download_chr1:
    output:
        protected("raw/chr1.fa")
    shadow: "shallow"
    run:
        shell("wget ftp://ftp.ensembl.org/pub/release-98/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.1.fa.gz"),
        shell("gunzip Homo_sapiens.GRCh38.dna.chromosome.1.fa.gz"),
        shell("cut -d ' ' -f 1 < Homo_sapiens.GRCh38.dna.chromosome.1.fa | sed 's/[nN]//g' >{output}")

#rule download_PaSGAL_data:
#    output:
#        protected("raw/PaSGAL/{file}")
#    shadow: "shallow"
#    run:
#        shell("wget {PaSGAL_DATA_URL}/{wildcards.file} -O {output}")

rule fa2prefix:
    input:
        "raw/{graph}.fa"
    output:
        "graphs/"+graph_params+".fa"
    run:
        if int(wildcards.head) > 0:
            shell("head -c {wildcards.head} {input} >{output}")
        else:
            shell("cp {input} {output}")
        #"echo -n 'S\t1\t' >{output} && tail -n +2 {input} | tr -d '\n' >>{output}"
        #"| head -c {wildcards.head} >>{output}"

#ruleorder generate_pasgal_graph > create_linear_graph

#rule generate_pasgal_graph:
#    input:
#        "raw/PaSGAL/{graph}.vg"
#    output:
#        vg = "graphs/graph{graph}_head0.vg"
#    shell:
#        "cp {input} {output.gfa}"

# make PaSGAL happy: add a fictive segment

rule fa2gfa_vg:
    input:
        "graphs/"+graph_params+".fa"
    output:
        gfa = "graphs/"+graph_params+".gfa",
        vg = "graphs/"+graph_params+".vg"
    run:
        if int(wildcards.head) > 0:
            shell("echo -n 'S\t1\t' >{output.gfa} && tail -n +2 {input} | tr -d '\n' >>{output.gfa} && echo -n '\nS\t2\tA\nL\t1\t+\t2\t+\t0M\n' >>{output.gfa}")
        else:
            shell("cp raw/{wildcards.graph}.gfa {output.gfa}")
        shell("{VG_BIN} view -Fv {output.gfa} >{output.vg}")

# art_illumina -ss MSv3 -i ecoli.fasta -l 20 -c 10000 -o tmp
#    "reads/"+params+"/reads.fq"

rule simulate_illumina_reads:
    input:
        "graphs/"+graph_params+".fa"
    params:
        prefix = "reads/"+graph_params+"_N{N}_m{m}_illumina_errors{errors}_seedlen{seedlen}/reads"
    output:
        "reads/"+graph_params+"_N{N}_m{m}_illumina_errors{errors}_seedlen{seedlen}/reads.fq"
    shadow: "shallow"
    shell:
        "{ART} -ss MSv3 -sam -i {input} -c {wildcards.N} -l {wildcards.m} -o {params.prefix} --rndSeed 42"

#rule simulate_ccs_reads:
#	input:
#		"graphs/"+graph_params+".fa"
#	params:
#		prefix = "reads/"+graph_params+"_N{N}_m{m}_ccs/reads"
#	output:
#		"reads/"+graph_params+"_N{N}_m{m}_ccs/reads.fq"
#	shadow: "shallow"
#	run:
#		shell("{PBSIM} --data-type CLR --depth 0.02 --model_qc pbsim_profiles/model_qc_clr --length-min {wildcards.m} --length-max {wildcards.m} --length-mean {wildcards.m} --accuracy-min 0.99 --accuracy-sd 0 --accuracy-mean 1 {input}")
#		#shell("{PBSIM} --hmm_model pbsim_models/P4C2.model --depth 0.001 --accuracy-min 0.99 --accuracy-max 0.99 --accuracy-mean 0.99 --accuracy-sd 0 --length-mean {wildcards.m} --length-sd 0 {input}")
#		#"{PBSIM} --sample-fastq pbsim_sample.fastq --depth 0.001 --accuracy-min 0.98 --accuracy-max 0.98 --length-mean {wildcards.m} --length-sd 0 {input}"
#		#"{PBSIM} --data-type CCS --depth 0.001 --model_qc pbsim_profiles/model_qc_ccs --length-mean {wildcards.m} --length-sd 0 {input}"
#		shell("cp -L sd_0001.fastq {params.prefix}.fq")
#		shell("cp -L sd_0001.maf {params.prefix}.maf")
#		shell("cp -L sd_0001.ref {params.prefix}.ref")

rule simulate_hifi_reads_fixed_len:
	input:
		"graphs/"+graph_params+".fa"
	params:
		prefix = "reads/"+graph_params+"_N{N}_m{m}_hifi_errors{errors}_seedlen{seedlen}/reads"
	output:
		"reads/"+graph_params+"_N{N}_m{m}_hifi_errors{errors}_seedlen{seedlen}/reads.fq"
	shadow: "shallow"
	shell:
		#"{RANDOMREADS} -Xmx1g build=1 ow=t seed=1 ref={input} illuminanames=t addslash=t pacbio=t pbmin={wildcards.errors} pbmax={wildcards.errors} reads={wildcards.N} paired=f gaussianlength=t minlength={wildcards.m} midlength={wildcards.m} maxlength={wildcards.m} out={output}"
		"{RANDOMREADS} -Xmx1g build=1 ow=t seed=1 ref={input} illuminanames=t addslash=t pacbio=t pbmin={wildcards.errors} pbmax={wildcards.errors} coverage={wildcards.N} paired=f gaussianlength=t minlength={wildcards.m} midlength={wildcards.m} maxlength={wildcards.m} out={output}"
		#"{RANDOMREADS} -Xmx1g build=1 ow=t seed=1 ref={input} illuminanames=t addslash=t pacbio=t pbmin={wildcards.errors} pbmax={wildcards.errors} coverage=$(echo 'scale=5; {wildcards.N}*1000000/{wildcards.head}' | bc) paired=f gaussianlength=t minlength={wildcards.m} midlength={wildcards.m} maxlength={wildcards.m} out={output}"
		#"{RANDOMREADS} -Xmx1g build=1 ow=t seed=1 ref={input} illuminanames=t addslash=t pacbio=t pbmin={wildcards.errors} pbmax={wildcards.errors} coverage=1 paired=f gaussianlength=t minlength={wildcards.m} midlength={wildcards.m} maxlength={wildcards.m} out={output}"

rule simulate_hifi_reads_natural:
	input:
		"graphs/"+graph_params+".fa"
	params:
		prefix = "reads/"+graph_params+"_N{N}_m{m}_hifi-natural_errors{errors}_seedlen{seedlen}/reads"
	output:
		"reads/"+graph_params+"_N{N}_{m}_hifi-natural_errors{errors}_seedlen{seedlen}/reads.fq"
	shadow: "shallow"
	shell:
		"{RANDOMREADS} -Xmx1g build=1 ow=t seed=1 ref={input} illuminanames=t addslash=t pacbio=t pbmin={wildcards.errors} pbmax={wildcards.errors} coverage=$(echo 'scale=5; {wildcards.N}*5000000/{wildcards.head}' | bc) paired=f gaussianlength=t minlength=5000 midlength=13000 maxlength=25000 out={output}"

rule head_reads:
    input:
        "raw/{reads_file}.fa"
    output:
        "reads/{reads_file}_reads{reads_num}.fa"
    shell:
        "head -n $(( 2 * {wildcards.reads_num} )) {input} >{output}"

rule head_pasgal_reads:
    input:
        "raw/PaSGAL/{reads_file}.fa"
    output:
        "reads/{reads_file}_reads{reads_num}.fa"
    shell:
        "head -n $(( 2 * {wildcards.reads_num} )) {input} >{output}"

